src/irmin-pack/irmin_pack_layers.ml

```diff
@@ -0,0 +1,765 @@
+          the other stores only need to update the flip. *)
+      let sync t =
+        let on_generation_change () =
+          Node.CA.clear_caches t.node;
+          Commit.CA.clear_caches t.commit
+        in
+        let on_generation_change_next_upper () =
+          Node.CA.clear_caches_next_upper t.node;
+          Commit.CA.clear_caches_next_upper t.commit
+        in
+        let flip =
+          Contents.CA.sync ~on_generation_change
+            ~on_generation_change_next_upper t.contents
+        in
+        t.flip <- flip;
+        Node.CA.update_flip ~flip t.node;
+        Commit.CA.update_flip ~flip t.commit;
+        Branch.update_flip ~flip t.branch
+
+      let clear t = Contents.CA.clear t.contents
+
+      (** migrate can be called on a layered store where only one layer exists
+          on disk. As migration fails on an empty store, we check which layer is
+          in the wrong version. *)
+      let migrate config =
+        if readonly config then raise RO_Not_Allowed;
+        let root = root config in
+        let root_upper1 = Filename.concat root (upper_root1 config) in
+        let config_upper1 = Conf.add config root_key (Some root_upper1) in
+        let root_upper0 = Filename.concat root (upper_root0 config) in
+        let config_upper0 = Conf.add config root_key (Some root_upper0) in
+        let root_lower = Filename.concat root (lower_root config) in
+        let config_lower = Conf.add config root_key (Some root_lower) in
+        [ root_upper1; root_upper0; root_lower ]
+        |> List.map (fun name ->
+               try
+                 let io =
+                   IO.v ~version:(Some current_version) ~fresh:false
+                     ~readonly:true (name // "store.pack")
+                 in
+                 (name, Some io)
+               with
+               | I.Invalid_version _ -> (name, None)
+               | e -> raise e)
+        |> List.fold_left
+             (fun to_migrate (name, io) ->
+               match io with
+               | None -> name :: to_migrate
+               | Some io ->
+                   IO.close io;
+                   to_migrate)
+             []
+        |> List.iter (fun root ->
+               if root = root_upper1 then migrate config_upper1;
+               if root = root_upper0 then migrate config_upper0;
+               if root = root_lower then migrate config_lower)
+
+      let layer_id t store_handler =
+        match store_handler with
+        | Commit_t k -> Commit.CA.layer_id t.commit k
+        | Node_t k -> Node.CA.layer_id t.node k
+        | Content_t k -> Contents.CA.layer_id t.contents k
+
+      let clear_previous_upper t =
+        Log.debug (fun l -> l "clear previous upper");
+        Contents.CA.clear_previous_upper t.contents >>= fun () ->
+        Node.CA.clear_caches_next_upper t.node;
+        Commit.CA.clear_caches_next_upper t.commit;
+        Branch.clear_previous_upper t.branch
```

Comment:
Hm, that would be nice, but I don't know how to do that. Can you point me to an example?

Reply:
It would first be necessary to define the common supertype of the four modules as its own module type (since `Branch` is an atomic write store while the others are not) – that may already exist. Then you can do something like:

```ocaml
let iter_lwt f { contents; node; commit; branch } =
  f (module Contents.CA) contents >>= fun () ->
  f (module Node.CA) node >>= fun () ->
  f (module Commit.CA) commit >>= fun () ->
  f (module Branch) branch

let dispatch t = function
 | Commit_t k -> f (module Commit.CA) t.commit k
 | Node_t k -> f (module Node.CA) t.node k
 | Content_t k -> f (module Contents.CA) t.contents k
```

etc. etc.

Reply:
Do you mind having a look [here](https://github.com/mirage/irmin/pull/882/commits/6804bf94b26f6fbcb83981e2f11ab977b46947d8#diff-52ce06d08638434eac31d8f0240b3b5eR286)? 

I can't figure out how to add `f` as an argument to `let iter_lwt f t =...` : I have to write its type explicitly, so that the type of the store is derived locally at each application of `f`.

Reply:
Hmm, yes. Sorry, my suggestion was insufficient. `f` needs a forall-quantifier, which then requires defining a record type:

```ocaml
        type 'a store_fn = {
          f : 't. (module S.STORE with type t = 't) -> 't -> 'a;
        }
        [@@ocaml.unboxed]

        let iter_lwt (f : unit Lwt.t store_fn) t : unit Lwt.t =
          f.f (module Contents) t.contents >>= fun () ->
          f.f (module Nodes) t.node >>= fun () ->
          f.f (module Commits) t.commit >>= fun () ->
          f.f (module Branch) t.branch
```

Sorry for wasting your time :slightly_smiling_face:

Reply:
That's neat! Thanks for helping me out :)